### PR TITLE
[KYUUBI #886] Add HTTP transport mode support to KYUUBI

### PR DESCRIPTION
**Pull request by SteNicholas**
### _Why are the changes needed?_

Add HTTP transport mode support to KYUUBI.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -857,15 +857,6 @@
                 <groupId>org.eclipse.jetty</groupId>
                 <artifactId>jetty-server</artifactId>
                 <version>${jetty.version}</version>
-                <exclusions>
-                    <!--
-                      Use `jakarta.servlet-api` instead.
-                      -->
-                    <exclusion>
-                        <groupId>javax.servlet</groupId>
-                        <artifactId>javax.servlet-api</artifactId>
```
Comment (pan3793) [inline pom.xml:866]: Do not remove this exclusion, `jakarta.servlet-api` is drop-in replacement of `javax.servlet-api`.

Comment (SteNicholas) [reply pom.xml:866]: @pan3793, in the `org.eclipse.jetty.server.Server` needs the `javax.servlet-api` denpendencies.

Comment (pan3793) [reply pom.xml:866]: I mean you can use `jakarta.servlet-api` to replace `javax.servlet-api`

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 2]
```diff
@@ -294,20 +296,28 @@ object KyuubiConf {
       .version("1.4.0")
       .fallbackConf(FRONTEND_BIND_HOST)
 
-  @deprecated(s"using ${FRONTEND_THRIFT_BINARY_BIND_PORT.key} instead", "1.4.0")
-  val FRONTEND_BIND_PORT: ConfigEntry[Int] = buildConf("frontend.bind.port")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:298]: let's keep this

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 3]
```diff
@@ -294,20 +296,28 @@ object KyuubiConf {
...
-      .fallbackConf(FRONTEND_BIND_PORT)
+      .intConf
+      .checkValue(p => p == 0 || (p > 1024 && p < 65535), "Invalid Port number")
+      .createWithDefault(10009)
+
+  val FRONTEND_THRIFT_HTTP_BIND_HOST: ConfigEntry[Option[String]] =
+    buildConf("frontend.thrift.http.bind.host")
+      .doc("Hostname or IP of the machine on which to run the thrift frontend service " +
+        "via http protocol.")
+      .version("1.5.0")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:311]: 1.6.0

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 4]
```diff
@@ -294,20 +296,28 @@ object KyuubiConf {
...
     buildConf("frontend.thrift.binary.bind.port")
       .doc("Port of the machine on which to run the thrift frontend service via binary protocol.")
       .version("1.4.0")
-      .fallbackConf(FRONTEND_BIND_PORT)
+      .intConf
+      .checkValue(p => p == 0 || (p > 1024 && p < 65535), "Invalid Port number")
+      .createWithDefault(10009)
+
+  val FRONTEND_THRIFT_HTTP_BIND_HOST: ConfigEntry[Option[String]] =
+    buildConf("frontend.thrift.http.bind.host")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:308]: Literally, we don't need a host config for every type of FrontendServices, different ports actually meet to goal alreadly

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 5]
```diff
@@ -294,20 +296,28 @@ object KyuubiConf {
...
+      .doc("Hostname or IP of the machine on which to run the thrift frontend service " +
+        "via http protocol.")
+      .version("1.5.0")
+      .fallbackConf(FRONTEND_BIND_HOST)
+
+  val FRONTEND_THRIFT_HTTP_BIND_PORT: ConfigEntry[Int] =
+    buildConf("frontend.thrift.http.bind.port")
+      .doc("Port of the machine on which to run the thrift frontend service via http protocol.")
+      .version("1.5.0")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:317]: 1.6.0

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 6]
```diff
@@ -364,6 +374,41 @@ object KyuubiConf {
       .version("1.4.0")
       .fallbackConf(FRONTEND_MAX_MESSAGE_SIZE)
 
+  val FRONTEND_THRIFT_HTTP_REQUEST_HEADER_SIZE: ConfigEntry[Int] =
+    buildConf("frontend.thrift.http.request.header.size")
+      .doc("Request header size in bytes, when using HTTP transport mode. Jetty defaults used.")
+      .version("1.5.0")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:380]: ditto

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 7]
```diff
@@ -364,6 +374,41 @@ object KyuubiConf {
...
+      .intConf
+      .createWithDefault(6 * 1024)
+
+  val FRONTEND_THRIFT_HTTP_RESPONSE_HEADER_SIZE: ConfigEntry[Int] =
+    buildConf("frontend.thrift.http.response.header.size")
+      .doc("Response header size in bytes, when using HTTP transport mode. Jetty defaults used.")
+      .version("1.5.0")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:387]: ditto

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 8]
```diff
@@ -364,6 +374,41 @@ object KyuubiConf {
...
+      .intConf
+      .createWithDefault(6 * 1024)
+
+  val FRONTEND_THRIFT_HTTP_MAX_IDLE_TIME: ConfigEntry[String] =
+    buildConf("frontend.thrift.http.max.idle.time")
+      .doc("Maximum idle time for a connection on the server when in HTTP mode.")
+      .version("1.5.0")
+      .stringConf
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:395]: timeConf

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 9]
```diff
@@ -364,6 +374,41 @@ object KyuubiConf {
...
+      .createWithDefault("1800s")
+
+  val FRONTEND_THRIFT_HTTP_PATH: ConfigEntry[String] =
+    buildConf("frontend.thrift.http.path")
+      .doc("Path component of URL endpoint when in HTTP mode.")
+      .version("1.5.0")
+      .stringConf
+      .createWithDefault("cliservice")
+
+  val FRONTEND_THRIFT_HTTP_COMPRESSION_ENABLED: ConfigEntry[Boolean] =
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:405]: we can remove this can compress all the time?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/THttpFrontendService.scala [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+import org.apache.kyuubi.{KyuubiException, Logging}
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.util.NamedThreadFactory
+
+/**
+ * Apache Thrift based hive service rpc
+ *  1. the server side implementation serves client-server rpc calls
+ *  2. the engine side implementations serve server-engine rpc calls
+ */
+abstract class THttpFrontendService(name: String)
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/THttpFrontendService.scala:43]: we don't need an abstract class in the common module, let's create a solid class in server module only

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/THttpFrontendService.scala [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+      }
+      server.foreach(_.join())
+    } catch {
+      case _: InterruptedException => error(s"$getName is interrupted")
+      case t: Throwable =>
+        error(s"Error starting $getName", t)
+        System.exit(-1)
+    }
+
+  override protected def stopServer(): Unit = {
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/THttpFrontendService.scala:147]: unnecessary

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/THttpFrontendService.scala [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+      case e: Throwable =>
+        error(e)
+        throw new KyuubiException(
+          s"Failed to initialize frontend service on $serverAddr:$portNum.",
+          e)
+    }
+    super.initialize(conf)
+  }
+
+  override def run(): Unit =
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/THttpFrontendService.scala:134]: unnecessary

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.kyuubi.ha.client.{KyuubiServiceDiscovery, ServiceDiscovery}
+import org.apache.kyuubi.service.{Serverable, Service, THttpFrontendService}
+import org.apache.kyuubi.service.TFrontendService.{CURRENT_SERVER_CONTEXT, OK_STATUS}
+import org.apache.kyuubi.session.KyuubiSessionImpl
+
+final class KyuubiTHttpFrontendService(
+    override val serverable: Serverable)
+  extends THttpFrontendService("KyuubiTHttpFrontend") {
+
+  override lazy val discoveryService: Option[Service] = {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:34]: let's make it None and create another PR for HA, as this may break something with binary

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    if (ServiceDiscovery.supportServiceDiscovery(conf)) {
+      Some(new KyuubiServiceDiscovery(this))
+    } else {
+      None
+    }
+  }
+
+  override def OpenSession(req: TOpenSessionReq): TOpenSessionResp = {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:42]: nothing changed from super, right?

**Review by yaooqinn — CHANGES_REQUESTED** [review]
We need to build a test for the newly added HTTP frontend

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 4 above)
Comment (SteNicholas) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:308]: @yaooqinn, I found every type of FrontendServices has the host config. Does every type of FrontendServices need to use the unfied host config?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 8 above)
Comment (SteNicholas) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:395]: @yaooqinn, this should be `stringConf`. I have handled the value of `frontend.thrift.http.max.idle.time` in `THttpFrontendService`.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 4 above)
Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:308]: not have a strong option on this, let's keep it

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 8 above)
Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:395]: we don't align with any configuration with HiveConf, we shall keep our time/interval configurations behave the same.

**Comment by SteNicholas** [comment]
@yaooqinn @pan3793, thanks for detailed review. I have addressed the above comments. Please help to take a look.

**Comment by yanghua** [comment]
@SteNicholas Check the CI?
